src/engine/Preprocessor.cpp

```
@@ -781,6 +794,278 @@ void Preprocessor::dumpAllBounds( const String &message )
+            : _weightedSumEquation( NULL )
+            , _activationFunction( NULL )
+        {
+        }
+
+        unsigned _weightedSumVariable;
+        unsigned _activationVariable;
+
+        PiecewiseLinearFunctionType _activationType;
+
+        const Equation *_weightedSumEquation;
+        const PiecewiseLinearConstraint *_activationFunction;
+    };
+
+    Map<Index, NeuronInformation> indexToNeuron;
+    Map<unsigned, Index> variableToIndex;
+    Map<unsigned, unsigned> layerToLayerSize;
+
+    ASSERT( !_preprocessed._networkLevelReasoner );
+
+    const List<Equation> &equations( _preprocessed.getEquations() );
+    const List<PiecewiseLinearConstraint *> plConstraints( _preprocessed.getPiecewiseLinearConstraints() );
+
+    // If the activation functions are not yet supported by the NLR,
+    // abort
+    for ( const auto constraint : plConstraints )
+        if ( !NetworkLevelReasoner::functionTypeSupported( constraint->getType() ) )
+            return false;
+
+    // Attempt to figure out the layered structure
+    List<unsigned> inputLayer;
+    List<unsigned> previousLayer;
+    List<unsigned> currentLayer;
+    List<unsigned> outputLayer;
+    unsigned currentLayerIndex = 1;
+
+    // The input and output layer should be defined in the input query
+    inputLayer = _preprocessed.getInputVariables();
+    previousLayer = inputLayer;
+    unsigned numVariablesHandledSoFar = inputLayer.size();
+    outputLayer = _preprocessed.getOutputVariables();
+    layerToLayerSize[0] = inputLayer.size();
+
+    // Store the input neurons in the topology
+    unsigned count = 0;
+    for ( const auto &neuron : inputLayer )
+    {
+        NeuronInformation neuronInfo;
+        neuronInfo._activationVariable = neuron;
+        Index index( 0, count );
+        indexToNeuron[index] = neuronInfo;
+        variableToIndex[neuron] = index;
+        ++count;
+    }
+
+    // Now, attempt to figure out the topology, layer by layer
+    while ( true )
```

Should we use the remaining number of nodes to identify in the loop condition? If we cannot identify new layers and we have non-zero remaining nodes we can report a reconstruction failure. Otherwise we can use that number to identify the output layer more reliably (i.e. when remainingNodes == outputLayer.size())...

---

There might be nodes not included in the topology, see above.